content/article/my-perl-wishlist-invariant-sigils-part-1.md

```diff
@@ -0,0 +1,181 @@
+| List of key/value pairs | `%z`: A hash, i.e., a list of key/value pair(s) | `@z%[0]` / `@z[0]:kv`: a list of the index `0` and the first element of `@z` | `%z%{0}` / `%z{0}:kv`: a list of the key `"0"` and the value of element `"0"` of `%z` |
+
+You can't always get what you want
+----------------------------------
+
+I prefer the adverb syntax.  It is easy to read, and it draws on
+all the expertise that has gone into the design of Raku.
+However, my preference has to be implementable.
+I'm not convinced that it is without major surgery.
+
+The Perl 5 parser decides how to interpret what is inside the brackets
+depending on the context provided by the slice.
+Perl interprets the `...` in `@foo[...]` as
+a list ([ref](https://github.com/Perl/perl5/blob/c58ad1f93e9ad7834d3735683462c07119aa87f5/perly.y#L1143-L1148)).
+In `$foo[...]`, Perl sees the `...` as a scalar expression
+([ref](https://github.com/Perl/perl5/blob/c58ad1f93e9ad7834d3735683462c07119aa87f5/perly.y#L958-L960)).
+As a result, for any slice syntax, the Perl 5 parser needs to know the desired
+type of result while parsing the subscript expression.  The adverb form,
+unfortunately, leaves the parser guessing until after the subscript
+is parsed.
+
+You can, in fact, hack the Perl parser to save the subscript
+until it sees a postfix adverb.  The parser can then apply the correct
+context.  I wrote a
+[proof-of-concept](https://github.com/Perl/perl5/compare/9786385e68f7f14df6f4dd0f04d2c72c0d9a2511...cxw42:3cd904788536b445c9c3abe9b469e1b569942051)
+for `@arr[expr]:v`.  It doesn't execute any code, but it does parse
+a postfix-adverb slice without crashing!  However, while writing that code,
+I ran across a surprise: new syntax isn't tied to a `use v5.xx`
+directive.
+
+It turns out the Perl parser lets code written against any Perl version
+use the latest syntax.  Both of the following command lines work on 5.30:
+
+```sh
+$ perl -Mstrict -Mwarnings -E 'my $z; $z->@* = 10..20'
+$ perl -Mstrict -Mwarnings -e 'my $z; $z->@* = 10..20'   # (!!!)
+```
+
+The second command line does not `use v5.30`, so you can't use `say`
+(introduced in v5.10).  However, you can use postderef (from v5.20)!
+As a result, a new syntax has to be meaningless in all previous Perl 5 versions.
+A postfix adverb fails this test. For example, the following is a valid
+Perl 5 program:
+
+```perl
+sub kv { "kv" }
+my @arr = 10..20;
+print 1 ? @arr[1,2]:kv;
+        # ^^^^^^^^^^^^ valid syntax, but not a slice :(
+print "\n";
+```
+
+Next Steps
+----------
+
+In Part 2, I will dig into the parser and tokenizer.  I will share some
+surprises I discovered while investigating postderef.  I will then describe
+a possible path forward.
+
+[]( vi: set ts=4 sts=4 sw=4 et ai tw=72 fo=2tbl: )
```

Comment:
In my browser this text shows up. If you change it to an html comment, it will be hidden.

Reply:
Removed